packages/repo/src/mst/mst.ts

```diff
@@ -766,14 +672,46 @@ export class MST implements DataStore {
   // Sync Protocol
 
   async writeToCarStream(car: BlockWriter): Promise<void> {
-    for await (const entry of this.walk()) {
-      if (entry.isTree()) {
-        const pointer = await entry.getPointer()
-        await this.storage.addToCar(car, pointer)
+    const entries = await this.getEntries()
+    const leaves = new CidSet()
+    let toFetch = new CidSet()
+    toFetch.add(await this.getPointer())
+    for (const entry of entries) {
+      if (entry.isLeaf()) {
+        leaves.add(entry.value)
       } else {
-        await this.storage.addToCar(car, entry.value)
+        toFetch.add(await entry.getPointer())
       }
     }
+    while (toFetch.size() > 0) {
+      const nextLayer = new CidSet()
+      const fetched = await this.storage.getBlocks(toFetch.toList())
+      if (fetched.missing.length > 0) {
+        throw new MissingBlocksError('mst node', fetched.missing)
+      }
+      for (const cid of toFetch.toList()) {
+        const found = await parse.getAndParse(fetched.blocks, cid, nodeDataDef)
+        await car.put({ cid, bytes: found.bytes })
+        const entries = await util.deserializeNodeData(this.storage, found.obj)
+
+        for (const entry of entries) {
+          if (entry.isLeaf()) {
+            leaves.add(entry.value)
+          } else {
+            nextLayer.add(await entry.getPointer())
+          }
+        }
+      }
+      toFetch = nextLayer
+    }
+    const leafData = await this.storage.getBlocks(leaves.toList())
+    if (leafData.missing.length > 0) {
+      throw new MissingBlocksError('mst leaf', leafData.missing)
+    }
+
+    for (const leaf of leafData.blocks.entries()) {
+      await car.put(leaf)
+    }
```

Comment:
Yup I think that's right^^ Getting them all at once is certainly better than what we were doing (involving hundreds/thousands of round trips), but there is definitely a better middle ground here

Reply:
Right on 👍

Followup question for my understanding— is the order of the blocks in the CAR significant?  I.e. is it important that the leaves are placed together last here?

Reply:
Hmmm I would say no 🤔 
I just did it here to reduce blockstore roundtrips. ie "get all leaves at once instead of making a roundtrip every time you encounter one"
I don't see a huge benefit to forcing order on a carfile. Maybe baking too many assumptions into it & I expect consumers to either hold the whole thing in memory or write it to disk. I suppose can see a benefit to having a deterministic layout in a carfile if only because then a carfile takes on an identity of it's own & you can know if you have the right thing by hash or something 🤔 
Filecoin uses depth-first search for a deterministic ordering: https://ipld.io/specs/transport/car/carv1/#determinism
I suspect ours might be breadth-first search if we do so 🤔